presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -1047,33 +1074,77 @@ private void logShuffleStatsSummary(ShuffleStatsKey key, List<PrestoSparkShuffle
                     DataSize.succinctBytes(bytesPerSecond));
         }
 
-        private static <T> void waitFor(Collection<Future<T>> futures)
-                throws SparkException
+        private long computeNextTimeout()
+                throws TimeoutException
         {
-            try {
-                for (Future<?> future : futures) {
-                    future.get();
+            long timeout = queryCompletionDeadline - System.currentTimeMillis();
+            if (timeout <= 0) {
+                throw new TimeoutException();
+            }
+            return timeout;
+        }
+    }
+
+    private static <T> void waitForActionsCompletionWithTimeout(Collection<JavaFutureAction<T>> actions, long timeout, TimeUnit timeUnit)
```

Comment:
Why `private static`,  we can always read `queryCompletionDeadline` directly, which is simpler and guaranteed to be accurate.  Parameters `timeout`, `timeUnit` can be calculated from `queryCompletionDeadline`  and current time.

Reply:
Same question :)

Reply:
How strongly do you feel about that? I just thought that with encapsulation it will be more readable. Basically the logic in the `PrestoSparkQueryExecution` is responsible for determining now much time we need to wait for the next operation to finish, and this method simply implements waiting.

Reply:
Maybe we can reuse this static function later